File: tests/interface/handler/role_test.go
package handler_test

import (
	"bytes"
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/go-chi/chi/v5"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/mock"

	roleusecase "go-clean-template/internal/application/usecase/role"
	"go-clean-template/internal/domain"
	"go-clean-template/internal/interface/dto"
	"go-clean-template/internal/interface/handler"
)

// ── mock types ───────────────────────────────────────────────────────────────

type MockRoleCreateUseCase struct{ mock.Mock }

func (m *MockRoleCreateUseCase) Execute(_ context.Context, req *roleusecase.CreateRoleInput) (*roleusecase.CreateRoleOutput, error) {
	args := m.Called(req)
	if args.Get(0) == nil {
		return nil, args.Error(1)
	}
	return args.Get(0).(*roleusecase.CreateRoleOutput), args.Error(1)
}

type MockRoleGetUseCase struct{ mock.Mock }

func (m *MockRoleGetUseCase) Execute(_ context.Context, req *roleusecase.GetRoleInput) (*roleusecase.GetRoleOutput, error) {
	args := m.Called(req)
	if args.Get(0) == nil {
		return nil, args.Error(1)
	}
	return args.Get(0).(*roleusecase.GetRoleOutput), args.Error(1)
}

type MockRoleListUseCase struct{ mock.Mock }

func (m *MockRoleListUseCase) Execute(_ context.Context, req *roleusecase.ListRolesInput) (*roleusecase.ListRolesOutput, error) {
	args := m.Called(req)
	if args.Get(0) == nil {
		return nil, args.Error(1)
	}
	return args.Get(0).(*roleusecase.ListRolesOutput), args.Error(1)
}

type MockRoleUpdateUseCase struct{ mock.Mock }

func (m *MockRoleUpdateUseCase) Execute(_ context.Context, id string, req *roleusecase.UpdateRoleInput) (*roleusecase.UpdateRoleOutput, error) {
	args := m.Called(id, req)
	if args.Get(0) == nil {
		return nil, args.Error(1)
	}
	return args.Get(0).(*roleusecase.UpdateRoleOutput), args.Error(1)
}

type MockRoleDeleteUseCase struct{ mock.Mock }

func (m *MockRoleDeleteUseCase) Execute(_ context.Context, req *roleusecase.DeleteRoleInput) error {
	args := m.Called(req)
	return args.Error(0)
}

// ── GetAll ───────────────────────────────────────────────────────────────────

func TestRoleGetAll_Success(t *testing.T) {
	mockUC := new(MockRoleListUseCase)
	h := handler.NewRoleHandler(nil, nil, mockUC, nil, nil)

	respData := &roleusecase.ListRolesOutput{
		Roles: []*domain.Role{
			{ID: "1", Name: "admin", Description: "Admin"},
		},
		Total:      1,
		Page:       1,
		PageSize:   20,
		TotalPages: 1,
	}
	mockUC.On("Execute", mock.AnythingOfType("*role.ListRolesInput")).Return(respData, nil)

	req := httptest.NewRequest("GET", "/api/v1/roles", nil)
	rec := httptest.NewRecorder()

	r := chi.NewRouter()
	r.Get("/api/v1/roles", h.GetAll)
	r.ServeHTTP(rec, req)

	assert.Equal(t, http.StatusOK, rec.Code)

	var resp dto.PaginatedResponse
	json.NewDecoder(rec.Body).Decode(&resp)
	assert.Equal(t, int64(1), resp.Total)
	assert.Equal(t, 1, resp.Page)
	assert.Equal(t, 20, resp.PageSize)
	assert.Equal(t, 1, resp.TotalPages)

	mockUC.AssertExpectations(t)
}

// ── GetByID ──────────────────────────────────────────────────────────────────

func TestRoleGetByID_Success(t *testing.T) {
	mockUC := new(MockRoleGetUseCase)
	h := handler.NewRoleHandler(nil, mockUC, nil, nil, nil)

	mockUC.On("Execute", &roleusecase.GetRoleInput{ID: "123"}).Return(
		&roleusecase.GetRoleOutput{Role: &domain.Role{ID: "123", Name: "admin", Description: "Admin"}}, nil,
	)

	req := httptest.NewRequest("GET", "/api/v1/roles/123", nil)
	rec := httptest.NewRecorder()

	r := chi.NewRouter()
	r.Get("/api/v1/roles/{id}", h.GetByID)
	r.ServeHTTP(rec, req)

	assert.Equal(t, http.StatusOK, rec.Code)

	var resp dto.RoleResponse
	json.NewDecoder(rec.Body).Decode(&resp)
	assert.Equal(t, "123", resp.ID)
	assert.Equal(t, "admin", resp.Name)

	mockUC.AssertExpectations(t)
}

func TestRoleGetByID_NotFound(t *testing.T) {
	mockUC := new(MockRoleGetUseCase)
	h := handler.NewRoleHandler(nil, mockUC, nil, nil, nil)

	mockUC.On("Execute", &roleusecase.GetRoleInput{ID: "999"}).Return(nil, domain.ErrNotFound)

	req := httptest.NewRequest("GET", "/api/v1/roles/999", nil)
	rec := httptest.NewRecorder()

	r := chi.NewRouter()
	r.Get("/api/v1/roles/{id}", h.GetByID)
	r.ServeHTTP(rec, req)

	assert.Equal(t, http.StatusNotFound, rec.Code)

	var resp map[string]string
	json.NewDecoder(rec.Body).Decode(&resp)
	assert.Equal(t, "not found", resp["error"])

	mockUC.AssertExpectations(t)
}

// ── Create ───────────────────────────────────────────────────────────────────

func TestRoleCreate_Success(t *testing.T) {
	mockUC := new(MockRoleCreateUseCase)
	h := handler.NewRoleHandler(mockUC, nil, nil, nil, nil)

	reqBody := `{"name":"admin","description":"Administrator role"}`
	mockUC.On("Execute", mock.AnythingOfType("*role.CreateRoleInput")).Return(
		&roleusecase.CreateRoleOutput{Role: &domain.Role{ID: "1", Name: "admin", Description: "Administrator role"}}, nil,
	)

	req := httptest.NewRequest("POST", "/api/v1/roles", bytes.NewBufferString(reqBody))
	req.Header.Set("Content-Type", "application/json")
	rec := httptest.NewRecorder()

	r := chi.NewRouter()
	r.Post("/api/v1/roles", h.Create)
	r.ServeHTTP(rec, req)

	assert.Equal(t, http.StatusCreated, rec.Code)

	var resp dto.RoleResponse
	json.NewDecoder(rec.Body).Decode(&resp)
	assert.Equal(t, "admin", resp.Name)

	mockUC.AssertExpectations(t)
}

func TestRoleCreate_InvalidBody(t *testing.T) {
	h := handler.NewRoleHandler(nil, nil, nil, nil, nil)

	req := httptest.NewRequest("POST", "/api/v1/roles", bytes.NewBufferString(`invalid json`))
	req.Header.Set("Content-Type", "application/json")
	rec := httptest.NewRecorder()

	r := chi.NewRouter()
	r.Post("/api/v1/roles", h.Create)
	r.ServeHTTP(rec, req)

	assert.Equal(t, http.StatusBadRequest, rec.Code)
}

func TestRoleCreate_AlreadyExists(t *testing.T) {
	mockUC := new(MockRoleCreateUseCase)
	h := handler.NewRoleHandler(mockUC, nil, nil, nil, nil)

	mockUC.On("Execute", mock.AnythingOfType("*role.CreateRoleInput")).Return(nil, domain.ErrAlreadyExists)

	req := httptest.NewRequest("POST", "/api/v1/roles", bytes.NewBufferString(`{"name":"admin","description":"Admin"}`))
	req.Header.Set("Content-Type", "application/json")
	rec := httptest.NewRecorder()

	r := chi.NewRouter()
	r.Post("/api/v1/roles", h.Create)
	r.ServeHTTP(rec, req)

	assert.Equal(t, http.StatusConflict, rec.Code)

	mockUC.AssertExpectations(t)
}

// ── Update ───────────────────────────────────────────────────────────────────

func TestRoleUpdate_Success(t *testing.T) {
	mockUC := new(MockRoleUpdateUseCase)
	h := handler.NewRoleHandler(nil, nil, nil, mockUC, nil)

	mockUC.On("Execute", "1", mock.AnythingOfType("*role.UpdateRoleInput")).Return(
		&roleusecase.UpdateRoleOutput{Role: &domain.Role{ID: "1", Name: "updated", Description: "Updated description"}}, nil,
	)

	req := httptest.NewRequest("PUT", "/api/v1/roles/1", bytes.NewBufferString(`{"name":"updated","description":"Updated description"}`))
	req.Header.Set("Content-Type", "application/json")
	rec := httptest.NewRecorder()

	r := chi.NewRouter()
	r.Put("/api/v1/roles/{id}", h.Update)
	r.ServeHTTP(rec, req)

	assert.Equal(t, http.StatusOK, rec.Code)

	var resp dto.RoleResponse
	json.NewDecoder(rec.Body).Decode(&resp)
	assert.Equal(t, "updated", resp.Name)

	mockUC.AssertExpectations(t)
}

// ── Delete ───────────────────────────────────────────────────────────────────

func TestRoleDelete_Success(t *testing.T) {
	mockUC := new(MockRoleDeleteUseCase)
	h := handler.NewRoleHandler(nil, nil, nil, nil, mockUC)

	mockUC.On("Execute", &roleusecase.DeleteRoleInput{ID: "1"}).Return(nil)

	req := httptest.NewRequest("DELETE", "/api/v1/roles/1", nil)
	rec := httptest.NewRecorder()

	r := chi.NewRouter()
	r.Delete("/api/v1/roles/{id}", h.Delete)
	r.ServeHTTP(rec, req)

	assert.Equal(t, http.StatusNoContent, rec.Code)

	mockUC.AssertExpectations(t)
}

func TestRoleDelete_NotFound(t *testing.T) {
	mockUC := new(MockRoleDeleteUseCase)
	h := handler.NewRoleHandler(nil, nil, nil, nil, mockUC)

	mockUC.On("Execute", &roleusecase.DeleteRoleInput{ID: "999"}).Return(domain.ErrNotFound)

	req := httptest.NewRequest("DELETE", "/api/v1/roles/999", nil)
	rec := httptest.NewRecorder()

	r := chi.NewRouter()
	r.Delete("/api/v1/roles/{id}", h.Delete)
	r.ServeHTTP(rec, req)

	assert.Equal(t, http.StatusNotFound, rec.Code)

	mockUC.AssertExpectations(t)
}

File: tests/handler/role_test.rs
use std::sync::Arc;

use axum::extract::{Path, State};
use axum::http::StatusCode;
use axum::response::IntoResponse;
use async_trait::async_trait;

use go_clean_template::application::usecase::role::*;
use go_clean_template::application::error::AppError;
use go_clean_template::domain;
use go_clean_template::interface::dto::CreateRoleRequest;
use go_clean_template::interface::handler::role::RoleHandler;

struct MockCreateRoleUC;
#[async_trait]
impl CreateRoleUseCase for MockCreateRoleUC {
    async fn execute(&self, input: CreateRoleInput) -> Result<CreateRoleOutput, AppError> {
        let role = domain::Role::new(input.name, input.description)?;
        Ok(CreateRoleOutput { role })
    }
}

struct MockGetRoleUC;
#[async_trait]
impl GetRoleUseCase for MockGetRoleUC {
    async fn execute(&self, input: GetRoleInput) -> Result<GetRoleOutput, AppError> {
        if input.id == "999" {
            return Err(AppError::NotFound);
        }
        let role = domain::Role::new("admin".into(), "Admin".into()).unwrap();
        Ok(GetRoleOutput { role })
    }
}

struct MockListRolesUC;
#[async_trait]
impl ListRolesUseCase for MockListRolesUC {
    async fn execute(&self, _input: ListRolesInput) -> Result<ListRolesOutput, AppError> {
        Ok(ListRolesOutput { roles: vec![], total: 0, page: 1, page_size: 20, total_pages: 1 })
    }
}

struct MockUpdateRoleUC;
#[async_trait]
impl UpdateRoleUseCase for MockUpdateRoleUC {
    async fn execute(&self, _id: &str, _input: UpdateRoleInput) -> Result<UpdateRoleOutput, AppError> {
        let role = domain::Role::new("admin".into(), "Admin".into()).unwrap();
        Ok(UpdateRoleOutput { role })
    }
}

struct MockDeleteRoleUC;
#[async_trait]
impl DeleteRoleUseCase for MockDeleteRoleUC {
    async fn execute(&self, _input: DeleteRoleInput) -> Result<(), AppError> { Ok(()) }
}

fn test_role_handler() -> RoleHandler {
    RoleHandler::new(
        Box::new(MockCreateRoleUC),
        Box::new(MockGetRoleUC),
        Box::new(MockListRolesUC),
        Box::new(MockUpdateRoleUC),
        Box::new(MockDeleteRoleUC),
    )
}

#[tokio::test]
async fn test_role_get_by_id_success() {
    let handler = Arc::new(test_role_handler());
    let response = RoleHandler::get_by_id(
        State(handler),
        Path("1".into()),
    ).await.into_response();
    assert_eq!(response.status(), StatusCode::OK);
}

#[tokio::test]
async fn test_role_get_by_id_not_found() {
    let handler = Arc::new(test_role_handler());
    let response = RoleHandler::get_by_id(
        State(handler),
        Path("999".into()),
    ).await.into_response();
    assert_eq!(response.status(), StatusCode::NOT_FOUND);
}

#[tokio::test]
async fn test_role_create_success() {
    let handler = Arc::new(test_role_handler());
    let response = RoleHandler::create(
        State(handler),
        axum::Json(CreateRoleRequest { name: "admin".into(), description: "Admin role".into() }),
    ).await.into_response();
    assert_eq!(response.status(), StatusCode::CREATED);
}

#[tokio::test]
async fn test_role_delete_success() {
    let handler = Arc::new(test_role_handler());
    let response = RoleHandler::delete(
        State(handler),
        Path("1".into()),
    ).await.into_response();
    assert_eq!(response.status(), StatusCode::NO_CONTENT);
}
